### DISCUSSION TITLE
Grain reconstruction of specific regions.

### DISCUSSION THREAD

**Discussion by ejivul35p** [mtex-toolbox/mtex#2077] [Ask Anything]
Hello,
I have encountered some issues with grain reconstruction in AA1100 aluminum alloy. Below are my EBSD image and the result of grain reconstruction using the FMC method.
![275-90-60-30-iq](https://github.com/mtex-toolbox/mtex/assets/156154684/0db6ff98-69b6-4d87-b3f1-29e8339a70e0)
![90-60-30-0 09](https://github.com/mtex-toolbox/mtex/assets/156154684/593dc083-a463-49cf-8b34-44fff9e97c09)

As shown in the EBSD image, there are many dark regions with poor signal quality. I would like to exclude these regions from analysis and only reconstruct the clear areas to calculate their grain sizes. Could you please advise me on how to reconstruct grains only in these specific areas by the  FMC method?

**Comment by kilir** [comment]
Hi,
one option might be to discard grains which are below a certain threshold of your signal quality after computing grains, See `grainMean` for that case. Another option would be to set the EBSD pixels below a given threshold to `'notIndexed'` prior to computing grains. E.g. `ebsd(ebsd.bc < 80) = 'notIndexed'` if bc would be your signal property.

Hope this helps.

Cheers,

Rüdiger

**Reply by ejivul35p** [reply]
Hello, thank you for your reply. I used the method "ebsd(ebsd.ci < 0.2) = 'notIndexed'" and obtained the EBSD image below.
![275-90-60-30-ci小於0 2](https://github.com/mtex-toolbox/mtex/assets/156154684/89eaee73-57a6-46a5-8e06-7a9620745be6)
However, when I proceeded to perform grain reconstruction using the FMC method, it still reconstructed the parts containing 'notIndexed' points.
![275-90-60-30-ci小於0 2-0 09](https://github.com/mtex-toolbox/mtex/assets/156154684/558998d9-d42a-4be6-9736-3ca561d50d21)
 Is it possible to set a condition: to remove a grain if it contains too many 'notIndexed' points after reconstruction by the FMC method?
Thank you for your help.

**Reply by kilir** [reply]
If the grains you show are all aluminum and none are `'notIndexed'`, could it be that you use `ebsd('indexed')` instead of `ebsd` in `calcGrains`?

Cheers,

Rüdiger

**Reply by ejivul35p** [reply]
Hello, thank you for your reply.Sorry for the late response.I have already tried to use this code:
 grains_FMC = calcGrains(ebsd('indexed'),'FMC',0.09)

 I've come up with a method to calculate the proportion of notindexed pixels in each grain after FMC reconstruction, and then remove the grain data with excessive proportions of notindexed pixels. I have successfully obtained the total number of pixels for each grain, but encountered difficulties in calculating the number of not indexed pixels in each grain. Here is my code:

ebsd = EBSD.load(fname,CS,'interface','csv');
plot(ebsd)
ebsd(ebsd.ci < 0.1) = 'notIndexed'
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'))
ebsd(grains(grains.grainSize < 2)) = [];
ebsd('notIndexed') = [];
plot(ebsd)

grains_FMC = calcGrains(ebsd('indexed'),'FMC',0.09)
grains = calcGrains(ebsd('indexed'))
grains_FMC = smooth(grains_FMC);
plot(grains_FMC.boundary,'linewidth',0.3)
plot(grains_FMC, grains_FMC.meanOrientation);

num_grains = numel(grains_FMC.id);
fprintf('晶粒数量：%d\n', num_grains);
grains_FMC(1017).grainSize

ebsd_grain_1017 = ebsd(grains_FMC(1017).id);
not_indexed_pixels_1017 = sum(strcmp(ebsd_grain_1017.id, ebsd('notIndexed')));
fprintf('未索引的像素数量：%d\n', not_indexed_pixels_1017);

I attempted to calculate how many notindexed pixels the grain.id=1017 contains. However, the answer appears to be 0, while I can visually confirm the presence of notindexed pixels in the region corresponding to grain.id=1017 in the EBSD image. Could you please advise me on how to modify my code to correctly calculate the number of notindexed pixels?Or how to calculate the number of indexed pixels in each grain?
![DF8CBE5F-EF36-40A5-9DF4-E648400859AB](https://github.com/mtex-toolbox/mtex/assets/156154684/f5913a9f-43ff-4ae5-9800-0344b8bb0363)

**Reply by Tom-G-Uni-Vienna** [reply]
Hello, 
I think what Rüdiger is getting at is that if (after setting poor quality points to 'notIndexed') you would instead use calcGrains(ebsd) not calcGrains(ebsd('indexed')), your not indexed points would be treated as a separate phase, instead of being ignored. You then get grains of the indexed phase, as well as grains of the not indexed phase.

calcGrains(ebsd('indexed')) is probably also the reason why your code doesn't find any not indexed pixels inside your grains - when you reconstructed your grains, the not indexed pixels were not part of the reconstruction process at all - therefore, their ids are not included as being part of the grains. 

Doing calcGrains(ebsd('indexed')) is the same as first doing ebsd('notIndexed') = [ ] and then calcGrains(ebsd) - the not indexed pixels have become "holes" in the data, rather than pixels that are present, but with phase set to notIndexed. Therefore, they are totally ignored (and absent).

If you do want to proceed by working out how much of a given grain is "holes", you could maybe compare the number of indexed pixels in each grain to the area of the grain. grains.grainSize gives the grain size in pixels, but it may be simply reporting the number of indexed pixels in the grain, not the area, I am not sure. What you could do is calculate the actual area that one pixel would take up (probably this can be done from the vertices of the unit cell saved in ebsd.unitCell), and afterwards do

Npixels_byarea = grains(1017).area / area_of_one_pixel

indexed_fraction = length(ebsd(grains(1017))) / Npixels_byarea

the latter command might be identical to the simpler sytax

indexed_fraction = grains(1017).grainSize / Npixels_byarea 

depending on what the actual definition of .grainSize is.

If you do use the grain area, make sure you do not smooth your grain boundaries first, for the most accurate result.

best regards,
Tom G
